site/book/05-developing-functions/02-developing-in-Go.md

```diff
@@ -13,121 +12,161 @@ the following features:
 
-Generate a Dockerfile for the function image:
+# Pass the ResourceList as StdIn to your function
+kpt fn source data --fn-config data/fn-config.yaml | go run main.go
 
-```shell
-$ go run ./main.go gen ./
+# Verify the KRM function behavior in the StdOutput `ResourceList`
 ```
 
-Build the image:
+### Build the KRM function as an executable
 
+After editing your function, you can release it as an executable
 ```shell
-$ docker build . -t gcr.io/project/fn-name:tag
+go build main.go -o ${FUNCTION_NAME}
+# Change permission mode if needed
+chmod -x ${FUNCTION_NAME}
 ```
 
-Optionally, push the image to a container registry:
-
+#### Run the executable imperatively via kpt
 ```shell
-$ docker push gcr.io/project/fn-name:tag
+export PATH=<a path to your KRM resources directory>
+export FN_PATH=<a path to your FunctionConfig .yaml file>
+
+kpt fn eval ${PATH} --exec ./${FUNCTION_NAME} --fn-config ${FN_PATH}
 ```
 
-Run the function imperatively as a container function:
+#### Run the executable declaratively via kpt
+
+You can add the function to `Kptfile` via `--save` flag
 
 ```shell
-$ kpt fn eval wordpress -i gcr.io/project/fn-name:tag -- myannotation=foo
+kpt fn eval ${PATH} --save --exec ./${FUNCTION_NAME} --fn-config ${FN_PATH}
+
+# Run the executable in the Kptfile pipeline
+kpt fn render --allow-exec ${PATH}
 ```
 
-## Next Steps
+Once you have a function binary that works, you can then proceed to
+creating the container image.
 
-- Read the package documentation:
+### Build the KRM function as a Docker image
 
-| Package                                    | Purpose               |
-| ------------------------------------------ | --------------------- |
-| [sigs.k8s.io/kustomize/kyaml/fn/framework] | Functions Framework   |
-| [sigs.k8s.io/kustomize/kyaml/yaml]         | Modify YAML resources |
+Generate a Dockerfile for the function image:
+```shell
+cat > Dockerfile << EOF 
+FROM golang:1.17-alpine3.15
+ENV CGO_ENABLED=0
+WORKDIR /go/src/
+COPY go.mod go.sum ./
+RUN go mod download
+COPY . .
+RUN go build -o /usr/local/bin/function ./
+FROM alpine:3.15
+COPY --from=0 /usr/local/bin/function /usr/local/bin/function
+ENTRYPOINT ["function"]
+EOF
+```
 
-- Take a look at the source code for [functions in the catalog] to better
-  understand how to develop functions in Go
+Build the image:
 
-[sigs.k8s.io/kustomize/kyaml/fn/framework]: https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml@v0.11.1/fn/framework#pkg-index
-[sigs.k8s.io/kustomize/kyaml/yaml]: https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml@v0.11.1/yaml
-[functions in the catalog]: https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/functions/go
+```shell
+export FN_CONTAINER_REGISTRY=<Your GCR or docker hub>
+$ docker build . -t ${FN_CONTAINER_REGISTRY}
+```
+
+To verify the image using the same `./data` resources
+```shell
+kpt fn eval ${PATH} --image ${FN_CONTAINER_REGISTRY}/${FUNCTION_NAME}:latest --fn-config ${FN_PATH}
```

Comment:
I'd suggest not to use `latest` tag in our tutorial. Maybe use `${TAG}` and set it as `my-tag`.